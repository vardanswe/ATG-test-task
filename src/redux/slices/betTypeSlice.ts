import { createSlice, createAsyncThunk, PayloadAction } from '@reduxjs/toolkit';
import { fetchProducts, fetchGame } from '../../utils/api';

interface Track {
    id: number;
    name: string;
}

export interface RaceStart {
    id: string;
    number: number;
    horse: {
        name: string;
        age: number;
        sex: string;
        trainer: {
            firstName: string;
            lastName: string;
        };
        pedigree: {
            father: {
                name: string
            };
        };

    };
    driver: {
        firstName: string;
        lastName: string;
    };
}

interface Race {
    starts: RaceStart[];
    id: string;
    name: string;
    startTime: string;
    number: number;
}

interface BetTypeState {
    selectedBetType: string | null;
    mostRecentResult: {
        betType: string;
        tracks: Track[];
        startTime: string;
    } | null;
    races: Race[];
    loading: boolean;
    error: string | null;
}

const initialState: BetTypeState = {
    selectedBetType: null,
    mostRecentResult: null,
    races: [],
    loading: false,
    error: null,
};

export const fetchBetTypeProducts = createAsyncThunk(
    'betType/fetchProducts',
    async (betType: string, { rejectWithValue }) => {
        try {
            const data = await fetchProducts(betType);
            // better to check by date but assuming first item in array is the most recent result for this test task only
            const mostRecentResult = data.results[0];
            const mostRecentId = mostRecentResult.id;
            // fetching game details using id from most recent record
            const gameDetails = await fetchGame(mostRecentId);
            const races = gameDetails.races;
            return { mostRecentResult, races };
        } catch (error: any) {
            return rejectWithValue(error.message);
        }
    }
);

const betTypeSlice = createSlice({
    name: 'betType',
    initialState,
    reducers: {
        setBetType: (state, action: PayloadAction<string>) => {
            state.selectedBetType = action.payload;
        },
    },
    extraReducers: (builder) => {
        builder
            .addCase(fetchBetTypeProducts.pending, (state) => {
                state.loading = true;
                state.error = null;
            })
            .addCase(fetchBetTypeProducts.fulfilled, (state, action) => {
                state.loading = false;
                state.mostRecentResult = action.payload.mostRecentResult;
                state.races = action.payload.races;
            })
            .addCase(fetchBetTypeProducts.rejected, (state, action) => {
                state.loading = false;
                state.error = action.payload as string;
            })
    },
});

export const { setBetType } = betTypeSlice.actions;
export default betTypeSlice.reducer;
