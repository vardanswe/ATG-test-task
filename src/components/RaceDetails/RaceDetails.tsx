import React, {useState} from 'react';
import {RaceStart} from '../../redux/slices/betTypeSlice';

interface RaceDetailsProps {
    raceStarts: RaceStart[];
}

// TODO brake down to small react components or use Storybook components

const RaceDetails: React.FC<RaceDetailsProps> = ({raceStarts}) => {
    const [expandedRow, setExpandedRow] = useState<string | null>(null);

    if (!raceStarts || raceStarts.length === 0) {
        return <p className="text-center text-gray-500">No race details available.</p>;
    }

    return (
        <div className="race-details bg-white shadow rounded-lg p-4">
            <table className="w-full table-auto border-collapse border border-gray-200">
                <thead className="bg-gray-100">
                <tr>
                    <th className="border border-gray-300 px-4 py-2 text-left text-gray-600">Number</th>
                    <th className="border border-gray-300 px-4 py-2 text-left text-gray-600">Horse Name</th>
                    <th className="border border-gray-300 px-4 py-2 text-left text-gray-600">Driver</th>
                </tr>
                </thead>
                <tbody>
                {raceStarts.map((start) => (
                    <React.Fragment key={start.id}>
                        <tr
                            onClick={() => setExpandedRow(expandedRow === start.id ? null : start.id)}
                            className={`cursor-pointer hover:bg-gray-50 ${
                                expandedRow === start.id ? 'bg-gray-100' : ''
                            }`}
                            aria-expanded={expandedRow === start.id ? 'true' : 'false'}
                        >
                            <td className="border border-gray-300 px-4 py-2">{start.number}</td>
                            <td className="border border-gray-300 px-4 py-2">{start.horse.name}</td>
                            <td className="border border-gray-300 px-4 py-2">
                                {start.driver.firstName} {start.driver.lastName}
                            </td>
                        </tr>
                        {expandedRow === start.id && (
                            <tr>
                                <td colSpan={3} className="border border-gray-300 px-4 py-2 bg-gray-50">
                                    <div className="p-4 space-y-4 bg-gray-100 rounded">
                                        <div className="flex flex-col">
                                            <span
                                                className="text-sm font-medium text-gray-600">Horse Trainer: {start.horse.trainer.firstName} {start.horse.trainer.lastName}</span>
                                        </div>
                                        <div className="flex flex-col">
                                            <span
                                                className="text-sm font-medium text-gray-600">
                                                Horse Father: {start.horse.pedigree.father.name}
                                            </span>
                                        </div>
                                    </div>
                                </td>
                            </tr>
                        )}
                    </React.Fragment>
                ))}
                </tbody>
            </table>
        </div>
    );
};

export default RaceDetails;
