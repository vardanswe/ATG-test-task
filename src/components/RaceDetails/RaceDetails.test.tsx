import React from 'react';
import {render, screen, fireEvent} from '@testing-library/react';
import RaceDetails from './RaceDetails';
import { RaceStart } from '../../redux/slices/betTypeSlice';

const mockRaceStarts: RaceStart[] = [
    {
        id: '1',
        number: 1,
        horse: {
            name: 'Lightning Bolt',
            age:4,
            sex: 'male',
            trainer: {
                firstName: 'John',
                lastName: 'Doe',
            },
            pedigree: {
                father: { name: 'Thunderstorm' },
            },
        },
        driver: { firstName: 'Jane', lastName: 'Smith' },
    },
    {
        id: '2',
        number: 2,
        horse: {
            name: 'Storm Chaser',
            age: 3,
            sex: 'female',
            trainer: {
                firstName: 'Emily',
                lastName: 'Blake',
            },
            pedigree: {
                father: { name: 'Hurricane' },
            },
        },
        driver: { firstName: 'Mike', lastName: 'Johnson' },
    },
];

describe('RaceDetails Component', () => {
    test('renders without crashing', () => {
        render(<RaceDetails raceStarts={[]} />);
        expect(screen.getByText(/No race details available./i)).toBeInTheDocument();
    });

    test('displays race start details', () => {
        render(<RaceDetails raceStarts={mockRaceStarts} />);
        expect(screen.getByText('Lightning Bolt')).toBeInTheDocument();
        expect(screen.getByText('Storm Chaser')).toBeInTheDocument();
        expect(screen.getByText('Jane Smith')).toBeInTheDocument();
        expect(screen.getByText('Mike Johnson')).toBeInTheDocument();
    });

    test('expands and collapses rows on click', () => {
        render(<RaceDetails raceStarts={mockRaceStarts} />);

        const firstRow = screen.getByText('Lightning Bolt');
        fireEvent.click(firstRow);

        // Check if expanded content is visible
        expect(screen.getByText(/Horse Trainer: John Doe/i)).toBeInTheDocument();
        expect(screen.getByText(/Horse Father: Thunderstorm/i)).toBeInTheDocument();

        // Collapse the row
        fireEvent.click(firstRow);
        expect(screen.queryByText(/Horse Trainer: John Doe/i)).not.toBeInTheDocument();
    });

    test('sets aria-expanded correctly', () => {
        render(<RaceDetails raceStarts={mockRaceStarts} />);

        const firstRow = screen.getByText('Lightning Bolt').closest('tr');
        expect(firstRow).toHaveAttribute('aria-expanded', 'false');

        fireEvent.click(firstRow!);
        expect(firstRow).toHaveAttribute('aria-expanded', 'true');

        fireEvent.click(firstRow!);
        expect(firstRow).toHaveAttribute('aria-expanded', 'false');
    });


    test('displays "No race details available" message when raceStarts is empty', () => {
        render(<RaceDetails raceStarts={[]} />);

        expect(screen.getByText('No race details available.')).toBeInTheDocument();
    });

    test('matches snapshot', () => {
        const { container } = render(<RaceDetails raceStarts={mockRaceStarts} />);
        expect(container).toMatchSnapshot();
    });
});
