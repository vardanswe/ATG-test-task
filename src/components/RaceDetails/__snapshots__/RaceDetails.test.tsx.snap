// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RaceDetails Component matches snapshot 1`] = `
<div>
  <div
    class="race-details bg-white shadow rounded-lg p-4"
  >
    <table
      class="w-full table-auto border-collapse border border-gray-200"
    >
      <thead
        class="bg-gray-100"
      >
        <tr>
          <th
            class="border border-gray-300 px-4 py-2 text-left text-gray-600"
          >
            Number
          </th>
          <th
            class="border border-gray-300 px-4 py-2 text-left text-gray-600"
          >
            Horse Name
          </th>
          <th
            class="border border-gray-300 px-4 py-2 text-left text-gray-600"
          >
            Driver
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          aria-expanded="false"
          class="cursor-pointer hover:bg-gray-50 "
        >
          <td
            class="border border-gray-300 px-4 py-2"
          >
            1
          </td>
          <td
            class="border border-gray-300 px-4 py-2"
          >
            Lightning Bolt
          </td>
          <td
            class="border border-gray-300 px-4 py-2"
          >
            Jane
             
            Smith
          </td>
        </tr>
        <tr
          aria-expanded="false"
          class="cursor-pointer hover:bg-gray-50 "
        >
          <td
            class="border border-gray-300 px-4 py-2"
          >
            2
          </td>
          <td
            class="border border-gray-300 px-4 py-2"
          >
            Storm Chaser
          </td>
          <td
            class="border border-gray-300 px-4 py-2"
          >
            Mike
             
            Johnson
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
